src/rpc/handlers/BookChanges.cpp

```diff
@@ -0,0 +1,250 @@
+#include <ripple/app/ledger/Ledger.h>
+#include <ripple/basics/ToString.h>
+
+#include <backend/BackendInterface.h>
+#include <rpc/RPCHelpers.h>
+
+#include <boost/json.hpp>
+#include <algorithm>
+
+namespace json = boost::json;
+using namespace ripple;
+
+namespace RPC {
+
+struct BookChange
+{
+    STAmount sideAVolume;
+    STAmount sideBVolume;
+    STAmount highRate;
+    STAmount lowRate;
+    STAmount openRate;
+    STAmount closeRate;
+};
+
+class BookChangesHandler
+{
+    std::reference_wrapper<Context const> context_;
+    std::map<std::string, BookChange> tally_;
+    std::optional<uint32_t> offerCancel_;
+
+public:
+    ~BookChangesHandler() = default;
+    explicit BookChangesHandler(Context const& context)
+        : context_{std::cref(context)}
+    {
+    }
+
+    BookChangesHandler(BookChangesHandler const&) = delete;
+    BookChangesHandler(BookChangesHandler&&) = delete;
+    BookChangesHandler&
+    operator=(BookChangesHandler const&) = delete;
+    BookChangesHandler&
+    operator=(BookChangesHandler&&) = delete;
+
+    /**
+     * @brief Handles the `book_change` request for given transactions
+     *
+     * @param transactions The transactions to compute changes for
+     * @return std::vector<BookChange> The changes
+     */
+    std::vector<BookChange>
+    handle(LedgerInfo const& ledger)
+    {
+        reset();
+
+        for (auto const transactions =
+                 context_.get().backend->fetchAllTransactionsInLedger(
+                     ledger.seq, context_.get().yield);
+             auto const& tx : transactions)
+        {
+            handleBookChange(tx);
+        }
+
+        std::vector<BookChange> changes;
+        std::transform(
+            std::make_move_iterator(std::begin(tally_)),
+            std::make_move_iterator(std::end(tally_)),
+            std::back_inserter(changes),
+            [](auto obj) { return obj.second; });
+        return changes;
+    }
+
+private:
+    inline void
+    reset() noexcept
+    {
+        tally_.clear();
+        offerCancel_ = std::nullopt;
+    }
+
+    void
+    handleAffectedNode(STObject const& node)
+    {
+        auto const& metaType = node.getFName();
+        auto const nodeType = node.getFieldU16(sfLedgerEntryType);
+
+        // we only care about ltOFFER objects being modified or
+        // deleted
+        if (nodeType != ltOFFER || metaType == sfCreatedNode)
+            return;
+
+        // if either FF or PF are missing we can't compute
+        // but generally these are cancelled rather than crossed
+        // so skipping them is consistent
+        if (!node.isFieldPresent(sfFinalFields) ||
+            !node.isFieldPresent(sfPreviousFields))
+            return;
+
+        auto& finalFields = (const_cast<STObject&>(node))
```

Comment:
Vetoing `const_cast`. 

Why do we need const cast here. There's got to be a way to do this without requiring this. Look in STObject.h in rippled to see if there's a way that doesn't require this.

Reply:
Oh god i did not even notice it tbh. - this was copied as is from the original implementation in `rippled`. I'll definitely try to find a way around this - thanks

Reply:
Yea it was. I'll submit a patch to `rippled` later - no worries.